storage-proofs/src/circuit/stacked/params.rs

```diff
@@ -82,21 +74,44 @@ impl<H: Hasher> Proof<H> {
             comm_d_leaf.clone(),
         )?;
 
+        let comm_r_last_data_leaf =
+            comm_r_last_proof.alloc_value(cs.namespace(|| "comm_r_last_data_leaf"))?;
+
         // verify encodings
-        for (layer, proof) in encoding_proofs.into_iter().enumerate() {
-            proof.synthesize(
-                cs.namespace(|| format!("encoding_proof_{}", layer)),
-                params,
-                replica_id,
-            )?;
+        for (i, proof) in encoding_proofs.into_iter().enumerate() {
+            let layer = i + 1;
+
+            if layer == layers {
+                proof.synthesize_decoded(
+                    cs.namespace(|| format!("encoding_proof_{}", layer)),
+                    params,
+                    replica_id,
+                    &comm_r_last_data_leaf,
+                    &comm_d_leaf,
+                )?;
+            } else {
+                let raw = replica_column_proof.c_x.get_node_at_layer(layer);
+                let encoded_node = num::AllocatedNum::alloc(
+                    cs.namespace(|| format!("enc_node_{}", layer)),
+                    || {
+                        raw.map(Into::into)
+                            .ok_or_else(|| SynthesisError::AssignmentMissing)
+                    },
+                )?;
+
+                proof.synthesize_key(
+                    cs.namespace(|| format!("encoding_proof_{}", layer)),
```

Comment:
Please don't call this `encoding_proof_…` if it's just the KDF. It could be `labeling_proof_…`.

Reply:
naming is hard <3

Reply:
will do a renaming pass at a later point to get this nice